### PR TITLE
Go result set

### PR DESCRIPTION
**Pull request by Aiee**
add wrapper for vertex, edge and path

Updates: tests added

### CODE REVIEW

File: go/client_test.go [hunk 1]
```diff
@@ -365,7 +359,7 @@ func TestReconnect(t *testing.T) {
 
 	// Send query to server periodically
 	for i := 0; i < timeoutConfig.MaxConnPoolSize; i++ {
-		time.Sleep(1 * time.Second)
+		time.Sleep(200 * time.Millisecond)
```
Comment (laura-ding) [inline go/client_test.go:358]: why the indentation is different ohters

File: go/result_set.go [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+		if err != nil {
+			return nil, err
+		}
+		valList = append(valList, val)
+	}
+	return valList, nil
+}
+
+func (res ResultSet) GetRowValues(index int) ([]*nebula.Value, error) {
+	return res.records[index].GetRowValue(index)
```
Comment (laura-ding) [inline go/result_set.go]: check the `index`

File: go/result_set.go [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+		edgeType:    edgeType,
+		edgeName:    name,
+		ranking:     ranking,
+		properties:  &properties,
+		keys:        keys,
+		values:      values,
+	}
+}
+
+func newPathWrapper(path *nebula.Path) *PathWrapper {
```
Comment (laura-ding) [inline go/result_set.go]: genPathWrapper

File: go/result_set.go [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+
+type Node struct {
+	vid           string
+	labels        []string
+	tagPropKeys   map[string][]string
+	tagPropValues map[string][]*nebula.Value
+}
+
+type Relationship struct {
+	srcVertexID string
```
Comment (laura-ding) [inline go/result_set.go]: vertexID is `Value` type.

File: go/result_set.go [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+type PathWrapper struct {
+	nodeList         []Node
+	relationshipList []Relationship
+	segments         []segment
+}
+
+func newResultSet(resp graph.ExecutionResponse) ResultSet {
+	var records []Record
+	var colNames []string
+	for _, name := range resp.Data.ColumnNames {
```
Comment (laura-ding) [inline go/result_set.go]: the `Data` can be null, add check

File: go/result_set.go [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+		colNames = append(colNames, string(name))
+	}
+	for _, row := range resp.Data.Rows {
```
Comment (laura-ding) [inline go/result_set.go]: ditto

File: go/result_set.go [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+	endNode      Node
+}
+
+type PathWrapper struct {
+	nodeList         []Node
+	relationshipList []Relationship
+	segments         []segment
+}
+
+func newResultSet(resp graph.ExecutionResponse) ResultSet {
```
Comment (laura-ding) [inline go/result_set.go]: modify to `genResultSet`

File: go/result_set.go [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+	}
+	return ResultSet{
+		errorCode:   resp.ErrorCode,
+		errorMsg:    resp.ErrorMsg,
+		columnNames: colNames,
+		records:     records,
+	}
+}
+
+func newNode(vertex *nebula.Vertex) *Node {
```
Comment (laura-ding) [inline go/result_set.go]: modify to `genNode`

File: go/result_set.go [hunk 9]
```diff
@@ -0,0 +1,528 @@
...
+
+	return &Node{
+		vid:           vid,
+		labels:        labels,
+		tagPropKeys:   tagPropKeys,
+		tagPropValues: tagPropValues,
+	}
+}
+
+func newRelationship(edge *nebula.Edge) *Relationship {
```
Comment (laura-ding) [inline go/result_set.go]: modify to `genRelationship`

File: go/result_set.go [hunk 10]
```diff
@@ -0,0 +1,594 @@
...
+}
+
+func isPath(value *nebula.Value) bool {
+	if value.IsSetPVal() {
+		return true
+	}
+	return false
+}
+
+func AreEqualRelationship(r1 Relationship, r2 Relationship) bool {
```
Comment (laura-ding) [inline go/result_set.go]: modify to `CompareRelationship`, and no need to define this function.. you can override the equal of the relationship. the same as the node and path

File: go/result_set.go [hunk 11]
```diff
@@ -0,0 +1,594 @@
...
+}
+
+func isEdge(value *nebula.Value) bool {
+	if value.IsSetEVal() {
+		return true
+	}
+	return false
+}
+
+func isPath(value *nebula.Value) bool {
```
Comment (laura-ding) [inline go/result_set.go]: What's the difference to directly use `value.IsSetPVal()`

File: go/result_set.go [hunk 12]
```diff
@@ -0,0 +1,594 @@
...
+	}
+	return nil, fmt.Errorf("Failed to get Value: requested coloumn name or index is invalid, the parameter shuold be int or string")
+}
+
+/*
+	Return a Node if the value at key is a vertex.
+	Key is column name being look up if key is a String.
+	If key is an int, key is the index of row.
+*/
+func (record Record) AsNode(key interface{}) (*Node, error) {
```
Comment (laura-ding) [inline go/result_set.go]: directly to define two functions.
and `the string param` function to call `int param function`

File: go/result_set.go [hunk 13]
```diff
@@ -0,0 +1,594 @@
...
+		if colName == name {
+			return index, nil
+		}
+		index++
+	}
+	return -1, fmt.Errorf("Column name does not exist")
+}
+
+// Return a node if the value at the given index of row is a Vertex
+func (record Record) getNodeByIndex(index int) (*Node, error) {
```
Comment (laura-ding) [inline go/result_set.go]: modify `getNodeByIndex(index int)` to `asNode(index int)`

File: go/result_set.go [hunk 14]
```diff
@@ -0,0 +1,594 @@
...
+	}
+	// Get vertex by row index
+	if reflect.TypeOf(key).String() == "int" {
+		index := key.(int)
+		return record.getRelationshipByIndex(index)
+	}
+	return nil, fmt.Errorf("Failed to get relationship: requested coloumn name or index is invalid")
+}
+
+func (record Record) getRelationshipByIndex(index int) (*Relationship, error) {
```
Comment (laura-ding) [inline go/result_set.go]: ditto

File: go/result_set.go [hunk 15]
```diff
@@ -0,0 +1,594 @@
...
+	}
+	// Get vertex by row index
+	if reflect.TypeOf(key).String() == "int" {
+		index := key.(int)
+		return record.getPathByIndex(index)
+	}
+	return nil, fmt.Errorf("Failed to get path: requested coloumn name or index is invalid")
+}
+
+func (record Record) getPathByIndex(index int) (*PathWrapper, error) {
```
Comment (laura-ding) [inline go/result_set.go]: ditto

File: go/result_set.go [hunk 16]
```diff
@@ -0,0 +1,594 @@
...
+func (relationship Relationship) Keys() []string {
+	return relationship.keys
+}
+
+func (relationship Relationship) Values() []*nebula.Value {
+	return relationship.values
+}
+
+func (path *PathWrapper) GetPathLength() int {
+	return len(path.nodeList) - 1
```
Comment (laura-ding) [inline go/result_set.go]: len(path.segments)

File: go/result_set.go [hunk 17]
```diff
@@ -0,0 +1,495 @@
...
+	index, err := record.getIndexbyColName(colName)
+	if err != nil {
+		return nil, err
+	}
+	return record._record[index], nil
+}
+
+func (record Record) getIndexbyColName(colName string) (int, error) {
+	index := 0
+	for _, name := range *record.columnNames {
```
Comment (laura-ding) [inline go/result_set.go]: use map to save the name and index

File: go/result_set.go [hunk 18]
```diff
@@ -0,0 +1,495 @@
...
+	}
+	index := node.nameIndexMap[tagName]
+	for k, v := range node.vertex.Tags[index].Props {
+		keyList = append(keyList, k)
+		valueList = append(valueList, &ValueWrapper{v})
+	}
+
+	for i := 0; i < len(keyList); i++ {
+		kvMap[keyList[i]] = valueList[i]
+	}
```
Comment (laura-ding) [inline go/result_set.go]: for k, v := range node.vertex.Tags[index].Props {
                kvMap[k] = &ValueWrapper{v}
		keyList = append(keyList, k)
	}

File: go/result_set.go [hunk 19]
```diff
@@ -0,0 +1,495 @@
...
+	for _, r := range path.relationshipList {
+		if r.IsEqualTo(relationship) {
+			return true
+		}
+	}
+	return false
+}
+
+func (path *PathWrapper) GetStartNode() *Node {
+	return path.segments[0].startNode
```
Comment (laura-ding) [inline go/result_set.go]: check segments

File: go/result_set.go [hunk 20]
```diff
@@ -0,0 +1,495 @@
...
+}
+
+func (path *PathWrapper) GetEndNode() *Node {
+	return path.segments[len(path.segments)].endNode
```
Comment (laura-ding) [inline go/result_set.go]: out of range

File: go/result_set_test.go [hunk 21]
```diff
@@ -0,0 +1,391 @@
...
+	"fmt"
+	"sort"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/vesoft-inc/nebula-clients/go/nebula"
+	"github.com/vesoft-inc/nebula-clients/go/nebula/graph"
+)
+
+func TestAsBool(t *testing.T) {
```
Comment (laura-ding) [inline go/result_set_test.go:33]: add testEmpty() and testNull()

File: go/value_wrapper.go [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+		return true
+	}
+	return false
+}
+
+func (valueWrapper ValueWrapper) AsBool() (bool, error) {
+	if valueWrapper.value.IsSetBVal() {
+		return valueWrapper.value.GetBVal(), nil
+	}
+	return false, fmt.Errorf("Failed to convert value to bool")
```
Comment (laura-ding) [inline go/value_wrapper.go]: add print the input value type

File: go/value_wrapper.go [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+
+type ValueWrapper struct {
+	value *nebula.Value
+}
+
+func (valueWrapper ValueWrapper) IsNull() bool {
+	if valueWrapper.value.IsSetNVal() {
+		return true
+	}
+	return false
```
Comment (yixinglu) [inline go/value_wrapper.go]:
```go
return valueWrapper.value.IsSetNVal()
```

File: go/value_wrapper.go [hunk 24]
```diff
@@ -233,11 +196,45 @@ func (valueWrapper ValueWrapper) AsRelationship() (*Relationship, error) {
...
+// Returns the value type of value in the valueWrapper in string
+func (valueWrapper ValueWrapper) GetType() string {
+	if valueWrapper.value.IsSetNVal() {
+		return "Null"
+	} else if valueWrapper.value.IsSetBVal() {
+		return "Bool"
+	} else if valueWrapper.value.IsSetIVal() {
+		return "Int"
+	} else if valueWrapper.value.IsSetFVal() {
+		return "float"
```
Comment (laura-ding) [inline go/value_wrapper.go:217]: Unified lowercase

**Review by laura-ding — DISMISSED** [review]
Well done. and you need to modify the session's return result and modify the example

File: go/result_set.go [hunk 25]
```diff
@@ -0,0 +1,496 @@
...
+func (res ResultSet) GetColNames() []string {
+	return res.columnNames
+}
+
+func (res ResultSet) GetErrorCode() graph.ErrorCode {
+	return res.resp.ErrorCode
+}
+
+func (res ResultSet) GetErrorMsg() []byte {
+	return res.resp.ErrorMsg
```
Comment (laura-ding) [inline go/result_set.go:235]: check the res.resp.ErrorMsg is no None

File: go/result_set.go [hunk 26]
```diff
@@ -0,0 +1,496 @@
...
+	return &Record{
+		columnNames:     &res.columnNames,
+		_record:         valWrap,
+		colNameIndexMap: &res.colNameIndexMap,
+	}, nil
+}
+
+// Returns all rows
+func (res ResultSet) GetRows() []*nebula.Row {
+	return res.resp.Data.Rows
```
Comment (laura-ding) [inline go/result_set.go:219]: check res.resp.Data is no None

**Review by yixinglu — APPROVED** [review]

**Review by laura-ding — APPROVED** [review]

**Pull request merged**